### DISCUSSION TITLE
What are the bootstrap steps to build a webapp with multiple diagram editors

### DISCUSSION THREAD

**Discussion by jmfrancois** [eclipse-glsp/glsp#840] [Q&A] [answered]
Hello,

I'm trying to understand the boostrap steps of a glsp client. I have mainly looked at the [workflow-standalone app](https://github.com/eclipse-glsp/glsp-client/tree/master/examples/workflow-standalone) which is the closest from my needs.
My use case: I want to build a webapp (SPA with react) where I have multiple item that can be edited. The user experience is close to a theia example in the end.

So from what I understand in the example we have:

* create a websocket object
* create a inversify container
* fill the container with all the mappings for injection (interface/symbol to implementation)
* create jsonrpc client over the websocket (I have seen implementation update but not released yet)
* diagram server init and connect using the jsonrpc client
* init a session
* dispatch action RequestModelAction

This is fine and It has been quite easy to adapt it and make it work for my use case.

But this is more hard to understand what can be shared accross the entire app and what is specific to an item (workflow in the example). Where is the line between the editor, the DI or the websocket. I want to be able to open multiple diagrams, so handle multiple div with each one a specific sprotty instance (custom id using overrideViewerOptions is working but clearly looks like a global thing).

It seems there is a one-one link between the inversify container and the session / model used. 
The only way I have found to make it works was to duplicate everything from the container with the di.config.ts per editor, same with websocket ....

I have looked quicly at theia where having multiple editors seems native but the amount of code make it hard to understand and debug and it uses socket and not websocket for glsp.

Could please explain me if I am on the wrong way to setup my app ? I didn't find or understand that part from the doc of glsp.

**Comment by tortmayr** [comment, marked as answer]
Hi @jmfrancois,

if you want to integrate GLSP in your own standalone webapp the `workflow-standalone-app` is definitely a good starting point.
However, you have to keep in mind that this application is mostly used for quick testing during development. As you already noticed it is not designed to support multiple GLSP diagrams.

I will try to quickly outline the necessary steps for bootstrapping a multi-diagram webapp using this figure:
![Untitled drawing](https://user-images.githubusercontent.com/2311075/210317807-e87fea30-8be9-48da-9b63-7b48ac3da7d7.png)


Typically an application has one global GLSP client. A GLSP client can handle multiple diagram sessions, whereas each session represents a single diagram instance. Each client is handled by exactly one GLSP server.
So in your case the GLSP client can be shared across the entire app. Whenever a new diagram is instantiated you have to create a new DI container and connect it with the GLSP client.

If you want  to  use the `workflow-standalone-app` as a baseline you have to make a minor adjustment in the [createContainer()](https://github.com/eclipse-glsp/glsp-client/blob/fe0b29761e9844d8541ce4cafeaf7e00976157e1/examples/workflow-standalone/src/di.config.ts#L21) utility function. You have to pass the `widgetId` as an argument instead of using a hardcoded `sprotty` id. The `widgetId` correlates with the id of the diagram base div. So (assuming the the corresponding div elements exists) each created DI container can now render into its on div.

```ts
export default function createContainer(widgetId:string): Container {
    const container = createWorkflowDiagramContainer(widgetId);
   ...
    return container;
}

```

With this change you now can setup the application as follows:

## On application start

1. Start the server executable (if its not already running)
2. Create a websocket object pointing to the GLSP server endpoint (e.g. `ws://localhost:5007/workflow`)
3. Create a create jsonrpc client over the websocket connection
4. Initialize the server and optionally cache the initialize result. (e.g. `const result= await client.initializeServer()`). The initialize result will be later needed to configure each individual diagram session.

## When you open a new diagram
1. Create a new div for your diagram. (e.g. with id `myDiagramWidget`)
2. Create a new diagram DI container (e.g. `createContainer("myDiagramWidget")`)
3. Retrieve the `GLSPDiagramServer` for the container and set a new clientId
4. Connect it with the global glsp client
5. Configure which actions should be sent to the server (`await configureServerActions(result, diagramType, container)`). If you don't have cached the initialize result you can simply call `const result= await client.initializeServer()` again to retrieve  it.
6. Initialize a session (`client.initializeClientSession()`)

7. Retrieve the `ActionDispatcher` from the container and dispatch the set of initial actions

**Reply by jmfrancois** [reply]
I tried and client.initializeServer call lead to an error:
```
Error: JsonrpcGLSPClient is not ready yet
    checkConnectionState base-jsonrpc-glsp-client.js:46
```

After looking at the code, the  start is called by the glsp-diagram-server during the connect(client)
```
await diagramServer.connect(glspClient);
```

So yes I can call the start myself, but it lead then to an error as the connect trigger the start.

So at the moment I can only do 1, 2 and 3 but not 4.

**Reply by tortmayr** [reply]
Ah I see,
in this case you can also just omit step 4. of `on application start`
and simply call `client.initializeServer` in step 5. (of `When you open a new diagram`)

**Reply by jmfrancois** [reply]
Or we can see this as a bug.

glspClient.start could checkConnectionState() before start it like other methods are doing. So start can be called multiple time.
Or diagramServer in the connect method could check the connection state before calling start.

**Reply by amiths89** [reply]
@tortmayr Am quite new to web technologies and front end so the client side is a bit confusing for me as of now, It would help if i can relate it to a code I am familiar with a little better. Would it be possible to explain the same thing using the glsp-example tasklist diagram (java-emf-theia template) when an additional diagram language is being added. I have mirrored the files used for the tasklist diagram for my new diagram and connected both to the same port 5007 and added the newly created module class to the same server module in the TasListDiagramLauncher. Am not sure where i have gone wrong but it looks like am creating a kind of ambiguity with both the diagrams.
Also i have mentioned the steps i followed in this discussion: https://github.com/eclipse-glsp/glsp/discussions/1285

**Reply by tortmayr** [reply]
Hi @amiths89 
I have created a simple demo for a second diagram language based on the java-emf-theia template:
https://github.com/eclipse-glsp/glsp-examples/tree/demo-multiple-diagrams
This introduces a second language `second-diagram` that is currently just a mirror of the `tasklist` example.
To test this simply copy the `example.tasklist` file and rename it to `example.second`. Than it should be openend with the `Second Diagram editor`
